depeche_db/_subscription.py

```diff
@@ -222,16 +238,64 @@ def _update_max_aggregated_stream_positions_cache(
             max_position, current_value
         )
 
+    def _get_cutoff(self, conn: SAConnection) -> int:
+        state = self._state_provider.read(self.name)
+        if self._cutoff_cache.calculated_at < _now() - _dt.timedelta(hours=2):
+            min_global_position = None
+            if self._stream.max_partition is not None:
+                all_partitions_in_state = all(
+                    n in state.positions for n in range(self._stream.max_partition + 1)
+                )
+                if all_partitions_in_state:
+                    vals = _sa.sql.Values(
+                        _sa.column("partition", _sa.Integer),
+                        _sa.column("position", _sa.Integer),
+                        name="vals",
+                    ).data(
+                        [
+                            (partition, position)
+                            for partition, position in state.positions.items()
+                        ]
+                    )
+                    vals_cte = _sa.select(vals.c.partition, vals.c.position).cte()
+                    qry = _sa.select(
+                        _sa.func.min(
+                            self._stream._table.c.origin_stream_global_position
+                        )
+                    ).select_from(
+                        vals_cte.join(
+                            self._stream._table,
+                            _sa.and_(
+                                vals_cte.c.partition == self._stream._table.c.partition,
+                                vals_cte.c.position == self._stream._table.c.position,
+                            ),
+                        )
+                    )
+                    min_global_position = conn.execute(qry).scalar_one_or_none()
+                    if min_global_position:
+                        cutoff = max(
+                            min_global_position
+                            - int(
+                                self._stream.get_hourly_message_rate(conn)
+                                * self._stream.lookback_for_gaps_hours
+                            ),
+                            -1,
```

Comment:
From the min global position, we go back (rate * lookback time) to account for possible new things on the aggregated stream with lower global positions caused by late/out of order transactions...